pkg/util/fxutil/provide_comp.go

```diff
@@ -0,0 +1,381 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2016-present Datadog, Inc.
+
+package fxutil
+
+import (
+	"errors"
+	"fmt"
+	"reflect"
+	"strings"
+
+	compDef "github.com/DataDog/datadog-agent/comp/def"
+	"go.uber.org/fx"
+)
+
+var (
+	compOutType = reflect.TypeOf((*compDef.Out)(nil)).Elem()
+	compLcType  = reflect.TypeOf((*compDef.Lifecycle)(nil)).Elem()
+	fxInType    = reflect.TypeOf((*fx.In)(nil)).Elem()
+	fxOutType   = reflect.TypeOf((*fx.Out)(nil)).Elem()
+	fxLcType    = reflect.TypeOf((*fx.Lifecycle)(nil)).Elem()
+)
+
+// ProvideComponentConstructor takes as input a Component constructor function
+// that uses plain (non-fx aware) structs as its argument and return value, and
+// returns an fx.Provide'd Option that will properly include that Component
+// into the fx constructor graph.
+//
+// For example, given:
+//
+//	type Provides struct {
+//	    My MyComponent
+//	}
+//	type Requires struct {
+//	    Dep MyDependency
+//	}
+//	func NewComponent(reqs Requires) Provides { ... }
+//
+// then:
+//
+//	ProvideComponentConstructor(NewComponent)
+//
+// will create these anonymous types:
+//
+//	type FxAwareProvides struct {
+//	    fx.Out
+//	    My MyComponent
+//	}
+//	type FxAwareRequires struct {
+//	    fx.In
+//	    Dep MyDependency
+//	}
+//
+// and essentially make this happen:
+//
+//	fx.Provide(func(fxr FxAwareRequires) fxp FxAwareProvides {
+//	    r := makePlainReqs(fxr)
+//	    p := NewComponent(r)
+//	    return makeFxAwareProvides(p)
+//	})
+func ProvideComponentConstructor(compCtorFunc interface{}) fx.Option {
+	inFxType, outFxType, outEmbeds, err := constructFxInAndOut(compCtorFunc)
```

Comment:
The whole PR and the code is great.
 
Because reflect is not easy to manipulate, I have tried to see if I can simplify a little the code and here is a first version without the support lifecycle and group value
```go
func createNewStructType[T any](newFields []reflect.StructField) reflect.Type {
	currentType := getType[T]()

	var allFields []reflect.StructField
	for i := 0; i < currentType.NumField(); i++ {
		allFields = append(allFields, currentType.Field(i))
	}
	allFields = append(allFields, newFields...)
	return reflect.StructOf(allFields)
}

// Create a new struct of type target and copy all fields from source to target based on the field name
func createStructFrom[T any](source T, target reflect.Type) reflect.Value {
	currentValue := reflect.ValueOf(source)

	newValue := reflect.New(target).Elem()
	for i := 0; i < target.NumField(); i++ {
		value := target.Field(i)
		if currentValue := currentValue.FieldByName(value.Name); currentValue.IsValid() {
			newValue.FieldByName(value.Name).Set(currentValue)
		}
	}
	return newValue
}

// Wrapper is a generic function that wraps a function and provides it as a dependency to the fx framework
func Wrapper[Input any, Output any](f func(Input) Output) fx.Option {
	inFxType := createNewStructType[Input]([]reflect.StructField{{Name: "In", Type: reflect.TypeOf(fx.In{}), Anonymous: true}})
	outFxType := createNewStructType[Output]([]reflect.StructField{{Name: "Out", Type: reflect.TypeOf(fx.Out{}), Anonymous: true}})
	funcFxType := reflect.FuncOf([]reflect.Type{inFxType}, []reflect.Type{outFxType}, false)
	fxAwareProviderFunc := reflect.MakeFunc(funcFxType, func(args []reflect.Value) []reflect.Value {
		input := createStructFrom(args[0].Interface(), getType[Input]())
		inputValue := input.Interface().(Input)
		output := f(inputValue)
		fxOutput := createStructFrom(output, outFxType)
		return []reflect.Value{fxOutput}
	})

	return fx.Provide(fxAwareProviderFunc.Interface())
}

func getType[T any]() reflect.Type {
	var v [0]T
	return reflect.TypeOf(v).Elem()
}
```
What do you think?

Reply:
I really like the simplification here. There's definitely ideas here that can be integrated into my original version. However, there's a couple of features that need adding that are going to make this code a lot more complicated, even aside from needing Lifecycle and Group Value support:
1) Error messages should be helpful and be handled gracefully. In this version if a user tries to pass an interface for `Input` the code will panic because the reflect code treats it as a struct. I agree we should drop support for using an interface for the input type, but it shouldn't panic.
2) We need to be able to have the return value from the constructor be either `Output` or `Output, error`. Since it's variadic we can't use the generic version, so we'll have to retain the runtime reflection checks on the function parameter.
3) I think we should also allow constructor to take zero arguments, since that's useful for components with no inputs. See https://github.com/DataDog/datadog-agent/pull/24836/files#diff-e51fe42b28e45a8598e7913950a5d252e7011a819038d5da090474258562af58R171

Reply:
@ogaca-dd Refactored the entirety of the implementation, taking a lot of inspiration from what you did in the simpler version, while still retaining many of the features we need.